packages/dds/sequence/src/revertibles.ts

```diff
@@ -0,0 +1,214 @@
+) {
+	const startSeg = interval.start.getSegment() as SharedStringSegment;
+	const endSeg = interval.end.getSegment() as SharedStringSegment;
+	const startRef = string.createLocalReferencePosition(
+		startSeg,
+		interval.start.getOffset(),
+		ReferenceType.StayOnRemove | ReferenceType.RangeBegin,
+		undefined,
+	);
+	const endRef = string.createLocalReferencePosition(
+		endSeg,
+		interval.end.getOffset(),
+		ReferenceType.StayOnRemove | ReferenceType.RangeEnd,
+		undefined,
+	);
+	revertibles.push({
+		event: IntervalEventType.DELETE,
+		interval,
+		start: startRef,
+		end: endRef,
+	});
+
+	return revertibles;
+}
+
+export function appendLocalChangeToRevertibles(
+	string: SharedString,
+	newInterval: SequenceInterval,
+	previousInterval: SequenceInterval,
+	revertibles: IntervalRevertible[],
+) {
+	const startSeg = previousInterval.start.getSegment() as SharedStringSegment;
+	const endSeg = previousInterval.end.getSegment() as SharedStringSegment;
+	const prevStartRef = string.createLocalReferencePosition(
+		startSeg,
+		previousInterval.start.getOffset(),
+		ReferenceType.StayOnRemove | ReferenceType.RangeBegin,
+		undefined,
+	);
+	const prevEndRef = string.createLocalReferencePosition(
+		endSeg,
+		previousInterval.end.getOffset(),
+		ReferenceType.StayOnRemove | ReferenceType.RangeEnd,
+		undefined,
+	);
+	revertibles.push({
+		event: IntervalEventType.CHANGE,
+		interval: newInterval,
+		start: prevStartRef,
+		end: prevEndRef,
+	});
+
+	return revertibles;
+}
+
+export function appendLocalPropertyChangedToRevertibles(
+	interval: SequenceInterval,
+	deltas: PropertySet,
+	revertibles: IntervalRevertible[],
+) {
+	revertibles.push({
+		event: IntervalEventType.PROPERTYCHANGED,
+		interval,
+		propertyDeltas: deltas,
+	});
+
+	return revertibles;
+}
+// Uses of referenceRangeLabels will be removed once AB#4081 is completed.
+function revertLocalAdd(
+	string: SharedString,
+	revertible: TypedRevertible<typeof IntervalEventType.ADD>,
+) {
+	const maybeId = revertible.interval.getIntervalId();
+	const id = idMap.get(maybeId) ?? maybeId;
+	const label = revertible.interval.properties.referenceRangeLabels[0];
+	string.getIntervalCollection(label).removeIntervalById(id);
+}
+
+function revertLocalDelete(
+	string: SharedString,
+	revertible: TypedRevertible<typeof IntervalEventType.DELETE>,
+) {
+	const label = revertible.interval.properties.referenceRangeLabels[0];
+	const start = string.localReferencePositionToPosition(revertible.start);
+	const end = string.localReferencePositionToPosition(revertible.end);
+	const type = revertible.interval.intervalType;
+	// reusing the id causes eventual consistency bugs, so it is removed here and recreated in add
+	const { intervalId, ...props } = revertible.interval.properties;
+	const int = string.getIntervalCollection(label).add(start, end, type, props);
+
+	idMap.forEach((oldId, newId) => {
+		if (int.getIntervalId() === oldId) {
+			idMap.set(intervalId, newId);
```

Comment:
This will never be true since `int` is newly created here. I think you mean:

```suggestion
		if (intervalId === newId) {
			idMap.set(oldId, int.getIntervalId());
```

Also add a test for this. E.g. move interval, remove interval, revert only the remove, remove interval, revert last remove + move.